File: src/main/scala/ll1rdl/ListLexer.scala
package ll1rdl

class ListLexer(input: String) extends Lexer(input) {
  override def getTokenName(tokenType: Int) =
    ListLexer.tokenNames(tokenType)

  override def nextToken(): Token = {
    while (c != Lexer.EOF) {
      c match {
        case ' ' | '\t' | '\n' | '\r' =>
          Ws()

        case ',' =>
          consume()
          return Token(ListLexer.COMMA, ",")

        case '[' =>
          consume()
          return Token(ListLexer.LBRACK, "[")

        case ']' =>
          consume()
          return Token(ListLexer.RBRACK, "]")

        case _ =>
          if (isLetter()) {
            return Name()
          } else {
            throw new Error(s"Invalid character: $c")
          }
      }
    }

    return Token(Lexer.EOF_TYPE, "<EOF>")
  }

  def isLetter() =
    c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z'

  def Ws() = {
    while (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
      consume();
    }
  }

  def Name(): Token = {
    val buff = new StringBuilder

    do {
      buff.append(c)
      consume()
    } while (isLetter())

    return Token(ListLexer.NAME, buff.toString())
  }
}

object ListLexer {
  val NAME = 2
  val COMMA = 3
  val LBRACK = 4
  val RBRACK = 5

  val tokenNames = Array(
    "n/a",
    "<EOF>",
    "NAME",
    "COMMA",
    "LBRACK",
    "RBRACK"
  )
}

File: ll1rdl/parsing/lexer/ListLexer.java
public class ListLexer extends Lexer {
  public static int NAME = 2;
  public static int COMMA = 2;
  public static int LBRACK = 2;
  public static int RBRACK = 2;

  public static String[] tokenNames = {
    "n/a",
    "<EOF>",
    "NAME",
    "COMMA",
    "LBRACK",
    "RBRACK"
  };

  public String getTokenName(int x) {
    return tokenNames[x];
  }

  public ListLexer(String input) {
    super(input);
  }

  boolean isLETTER() {
    return c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z';
  }

  public Token nextToken() {
    while (c != EOF) {
      switch (c) {
        case ' ':
        case '\t':
        case '\n':
        case '\r':
          WS();
          continue;

        case ',':
          consume();
          return new Token(COMMA, ",");

        case '[':
          consume();
          return new Token(LBRACK, "[");

        case ']':
          consume();
          return new Token(RBRACK, "]");

        default:
          if (isLETTER()) {
            return NAME();
          }

          throw new Error("Invalid character: " + c);
      }
    }

    return new Token(EOF_TYPE, "<EOF>");
  }

  void WS() {
    while (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
      consume();
    }
  }

  Token NAME() {
    StringBuilder buff = new StringBuilder();

    do {
      buff.append(c);
      consume();
    } while (isLETTER());

    return new Token(NAME, buff.toString());
  }
}
